File: backend/app/models/lease.py
"""Lease domain model for tenant agreements.

The Lease model tracks tenant agreements with embedded recovery profile (JSONB)
containing all terms needed for the calculation engine.
"""

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .enums import LeaseStatus
from .lease_recovery_profile import LeaseRecoveryProfile, LeaseRecoveryProfileUpdate


class Lease(BaseModel):
    """Full lease model from database.

    Contains all lease data including the embedded recovery profile
    used by the financial calculation engine.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    property_id: UUID
    unit_id: UUID | None = None
    tenant_name: str = Field(..., min_length=1, max_length=255)
    start_date: date
    end_date: date
    status: LeaseStatus = Field(default=LeaseStatus.DRAFT)
    recovery_profile: LeaseRecoveryProfile
    document_url: str | None = Field(
        None, max_length=2048, description="S3 link to lease PDF"
    )
    created_at: datetime
    updated_at: datetime

    @model_validator(mode="after")
    def validate_dates(self) -> "Lease":
        """Ensure end_date is after start_date."""
        if self.end_date <= self.start_date:
            raise ValueError("End date must be after start date")
        return self


class LeaseCreate(BaseModel):
    """DTO for creating a lease.

    Requires property_id and all base fields. Unit_id is optional
    since a lease might cover multiple units or the entire property.
    """

    property_id: UUID
    unit_id: UUID | None = None
    tenant_name: str = Field(..., min_length=1, max_length=255)
    start_date: date
    end_date: date
    status: LeaseStatus = Field(default=LeaseStatus.DRAFT)
    recovery_profile: LeaseRecoveryProfile
    document_url: str | None = Field(None, max_length=2048)

    @model_validator(mode="after")
    def validate_dates(self) -> "LeaseCreate":
        """Ensure end_date is after start_date."""
        if self.end_date <= self.start_date:
            raise ValueError("End date must be after start date")
        return self


class LeaseUpdate(BaseModel):
    """DTO for updating a lease.

    All fields are optional for partial updates. Date validation
    must be checked at the service layer when merging with existing values.
    """

    tenant_name: str | None = Field(None, min_length=1, max_length=255)
    start_date: date | None = None
    end_date: date | None = None
    status: LeaseStatus | None = None
    recovery_profile: LeaseRecoveryProfileUpdate | None = None
    unit_id: UUID | None = None
    document_url: str | None = Field(None, max_length=2048)


class LeaseSummary(BaseModel):
    """Summary view of a lease for list displays.

    Contains essential fields without the full recovery profile.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    property_id: UUID
    unit_id: UUID | None = None
    tenant_name: str
    start_date: date
    end_date: date
    status: LeaseStatus

File: frontend/src/types/lease.ts
/**
 * Lease domain types for tenant agreements.
 *
 * These Zod schemas match exactly with backend/app/models/lease.py.
 * The Lease model tracks tenant agreements with embedded recovery profile.
 */

import { z } from 'zod'

import { LeaseStatus } from './enums'
import {
  LeaseRecoveryProfileSchema,
  LeaseRecoveryProfileUpdateSchema,
} from './lease-recovery-profile'

/**
 * Zod schema for LeaseStatus enum values.
 */
export const LeaseStatusSchema = z.enum([
  'draft',
  'active',
  'expired',
  'terminated',
])

/**
 * Full Lease model stored in database.
 *
 * Contains all lease data including the embedded recovery profile
 * used by the financial calculation engine.
 */
export const LeaseSchema = z
  .object({
    id: z.string().uuid(),
    property_id: z.string().uuid(),
    unit_id: z.string().uuid().nullable().optional(),
    tenant_name: z.string().min(1).max(255),
    start_date: z.string().date(),
    end_date: z.string().date(),
    status: LeaseStatusSchema.default('draft'),
    recovery_profile: LeaseRecoveryProfileSchema,
    document_url: z.string().max(2048).nullable().optional(),
    created_at: z.string().datetime(),
    updated_at: z.string().datetime(),
  })
  .refine((data) => new Date(data.end_date) > new Date(data.start_date), {
    message: 'End date must be after start date',
    path: ['end_date'],
  })

export type Lease = z.infer<typeof LeaseSchema>

/**
 * DTO for creating a lease.
 *
 * Requires property_id and all base fields. Unit_id is optional
 * since a lease might cover multiple units or the entire property.
 */
export const LeaseCreateSchema = z
  .object({
    property_id: z.string().uuid(),
    unit_id: z.string().uuid().nullable().optional(),
    tenant_name: z.string().min(1).max(255),
    start_date: z.string().date(),
    end_date: z.string().date(),
    status: LeaseStatusSchema.default('draft'),
    recovery_profile: LeaseRecoveryProfileSchema,
    document_url: z.string().max(2048).nullable().optional(),
  })
  .refine((data) => new Date(data.end_date) > new Date(data.start_date), {
    message: 'End date must be after start date',
    path: ['end_date'],
  })

export type LeaseCreate = z.infer<typeof LeaseCreateSchema>

/**
 * DTO for updating a lease.
 *
 * All fields are optional for partial updates. Date validation
 * must be checked at the service layer when merging with existing values.
 */
export const LeaseUpdateSchema = z.object({
  tenant_name: z.string().min(1).max(255).optional(),
  start_date: z.string().date().optional(),
  end_date: z.string().date().optional(),
  status: LeaseStatusSchema.optional(),
  recovery_profile: LeaseRecoveryProfileUpdateSchema.optional(),
  unit_id: z.string().uuid().nullable().optional(),
  document_url: z.string().max(2048).nullable().optional(),
})

export type LeaseUpdate = z.infer<typeof LeaseUpdateSchema>

/**
 * Summary view of a lease for list displays.
 *
 * Contains essential fields without the full recovery profile.
 */
export const LeaseSummarySchema = z.object({
  id: z.string().uuid(),
  property_id: z.string().uuid(),
  unit_id: z.string().uuid().nullable().optional(),
  tenant_name: z.string().min(1).max(255),
  start_date: z.string().date(),
  end_date: z.string().date(),
  status: LeaseStatusSchema,
})

export type LeaseSummary = z.infer<typeof LeaseSummarySchema>

/**
 * Helper to check if a lease status value is valid.
 */
export const isValidLeaseStatus = (value: string): value is LeaseStatus => {
  return Object.values(LeaseStatus).includes(value as LeaseStatus)
}
